src/graph_helper.rs

```diff
@@ -0,0 +1,507 @@
+    /// compile.
+    pub fn needs_compile(&self) -> bool {
+        self.needs_compile
+    }
+
+    /// Returns `true` if any nodes have had the state of their edges
+    /// (port connections) changed since the last call to
+    /// `AudioGraphHelper::nodes_with_dirty_edges()`.
+    ///
+    /// If `true`, then you can call
+    /// `AudioGraphHelper::nodes_with_dirty_edges()` to get the list of
+    /// these nodes, and then call `AudioGraphHelper::node_edges()` on
+    /// each one to retrieve its edges.
+    pub fn has_nodes_with_dirty_edges(&self) -> bool {
+        !self.nodes_with_dirty_edges.is_empty()
+    }
+
+    /// Returns a list of all nodes that have had the state of their
+    /// edges (port connections) changed since the last call to this
+    /// method.
+    ///
+    /// Call `AudioGraphHelper::node_edges()` on each node to retrieve
+    /// its edges.
+    pub fn nodes_with_dirty_edges(&mut self) -> Vec<NodeID> {
+        self.nodes_with_dirty_edges.drain().collect()
+    }
+
+    /// The total number of port types that can exist in this audio
+    /// graph. For example, if your graph can have an audio port type
+    /// and an event port type, then this should be `2`.
+    ///
+    /// Ports of different types cannot be connected together.
+    pub fn num_port_types(&self) -> usize {
+        self.num_port_types
+    }
+
+    fn remove_edges_with_port(&mut self, node_id: NodeID, port_id: PortID) {
+        let mut edges_to_remove: Vec<EdgeID> = Vec::new();
+
+        // Remove all existing edges which have this port.
+        for edge in self.edges.values() {
+            if (edge.src_node == node_id && edge.src_port == port_id)
+                || (edge.dst_node == node_id && edge.dst_port == port_id)
+            {
+                edges_to_remove.push(edge.id);
+            }
+        }
+
+        for edge_id in edges_to_remove.iter() {
+            let edge = self.edges.remove(edge_id).unwrap();
+
+            self.node_edges
+                .get_mut(&edge.src_node)
+                .unwrap()
+                .remove_outgoing(edge.id);
+            self.node_edges
+                .get_mut(&edge.dst_node)
+                .unwrap()
+                .remove_incoming(edge.id);
+
+            self.nodes_with_dirty_edges.insert(edge.src_node);
+            self.nodes_with_dirty_edges.insert(edge.dst_node);
+        }
+    }
+
+    fn cycle_detected(&self) -> bool {
+        let graph_ir = GraphIR::start(self.num_port_types, &self.nodes, &self.node_edges);
+        graph_ir.tarjan() > 0
+    }
+}
+
+/// The edges (port connections) that exist on a given [Node].
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone)]
+pub struct NodeEdges {
+    /// The edges connected to this node's input ports.
+    pub incoming: Vec<Edge>,
+    /// The edges connected to this node's output ports.
+    pub outgoing: Vec<Edge>,
+}
```

Comment:
One of the goals of the new architecture was to avoid keeping this state outside of the GraphIR, so the caller would only need to maintain the list of node IDs and their edges. 

I think this should be moved to the GraphIR and the logic for cycle checking added there as well.

Reply:
Additionally, you can use just a single `Vec<Edge>` here since `Edge`s contain their directional information. You can add a helper function to return `incoming` or `outgoing` from the `NodeEdges` data structure.

Reply:
> Additionally, you can use just a single `Vec<Edge>` here since `Edge`s contain their directional information. You can add a helper function to return `incoming` or `outgoing` from the `NodeEdges` data structure.

Like I mentioned above, every place this map is used in the algorithm, it is only iterating over the incoming or the outgoing edges. Filtering after the fact is less efficient.

Reply:
> One of the goals of the new architecture was to avoid keeping this state outside of the GraphIR, so the caller would only need to maintain the list of node IDs and their edges.
> 
> I think this should be moved to the GraphIR and the logic for cycle checking added there as well.

So basically what I want is there to be two ways to use this crate: One where you just pass your list of nodes and edges into the `compile()` method, and one where you use the helper struct to more optimally persist state, especially if you plan on constantly connecting and disconnecting edges.

Reply:
I also think the `GraphIR` struct should be private to the crate.

Reply:
I think of the GraphIR as the public structure that callers hold around, and it contains all the graph associated logic. 

I think a mutable/imperative API on top is valuable, but it should translate into a select set of calls to the graph IR. 

You can even add a bypass parameter to skip cycle checking in the main compile driver, so callers can choose if they want to uphold the invariants or not.